neuralNetwork.py

```diff
@@ -53,22 +97,116 @@ def calculate_cost_funcion(self, test_set):
             for i in range(len(f_k)):
-                cummulative_sum += (-y_k[i] * log(f_k[i])) - ((1-y_k[i]) * log(1-f_k[1]))
-            cost_function += cummulative_sum
+                cummulative_sum += (-y_k[i] * np.log(f_k[i])) - ((1-y_k[i]) * np.log(1-f_k[i]))
 
+            cost_function += cummulative_sum[0]
+            
+        cost_function = cost_function/n
+      
         theta_sum = 0
         for i in self.theta_list:
             theta_sum += i.sum() # cada elemento da lista de thetas é uma np.matrix
 
         cost_function += (self.reg_factor / 2*n) * theta_sum
-
+        
         return cost_function
+
+
+    def backpropagation(self, test_set, debug = False):
+        '''
+            Atualiza os pesos da rede neural com base nas instâncias de treino
+
+            1. calcular os deltas de cada neurônio:
+                para a camada de saída: delta = (f_x - y)
+                para as camadas ocultas: delta = sum(theta_(k+1)*delta_(k+1))*ativação_k*(1-ativação_k), onde k é a camada atual
+            2. calcular os gradientes dos pesos:
+                d/d(theta_ijk) = ativação_jk*delta_ik+1, onde theta_ijk é o peso entre o neurÔnio j na camada k e o neurônio i na camada k+1
+            3. atualizar cada peso com o valor do gradiente:
+                theta_ijk = theta_ijk -learning_rate*d/d(theta_ijk)
+        '''
+        L = len(self.activation)
+
+
+        for _,entry in test_set.iterrows():
+            y_k = []
+            for i in self.output_columns:
+                y_k.append(entry[i])
+            f_k = self.predict(entry.drop(self.output_columns))
+
+            #1. calcular os deltas de cada neurônio:
+            deltas = [None]*L
+
+            # caso especial da camada de saída
+            deltas[L - 1] = f_k - y_k
+
+            if debug:
+                saida = ['%.5f' % n for n in deltas[L-1]]
+                print('\t\tdelta'+str(L) + ': [' + ' '.join(saida) + ']')
+
+            # resto das camadas
+            for i in reversed(range(1,L - 1)): # primeira camada não tem ativação
+                deltas[i] = []
+                #print('theta_list[i]: '+str(self.theta_list[i]))
+                first = deltas[i+1].T.dot(self.theta_list[i])
+          
+                second = np.subtract(self.activation[i], np.square(self.activation[i]))
+
+                deltas[i] =  np.multiply(first.T,second) # element-wise multiplication
+                if debug:
+                    #print('deltas[i]: ' +str(deltas[i]))
+                    print_camada = ['%.5f' % n for n in deltas[i]]
+                    print('\t\tdelta'+str(i+1) + ': [' + ' '.join(print_camada) + ']')
+
+
+            #2. calcular os gradientes dos pesos:
+            gradients = [None] * (L-1)
+            bias_gradients = [None]*(L-1) # n sei se é L
+            for i in reversed(range(len(self.theta_list))):
+                gradients[i] = self.activation[i].dot(deltas[i+1].T).T + self.reg_factor*self.theta_list[i]
```

Comment:
esse warning tá fritando meu cérebro 
`VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray `

Reply:
É do numpy na linha 75 isso? Não apareceu pra mim, talvez eu precise atualizar a versão do numpy